test/model/autocomplete_test.dart

```diff
@@ -274,35 +276,183 @@ void main() {
+        await Future(() {});
+        expect(view.results);
+      }
+    }
+
+    List<User> generateUsers({required int count}) {
+      return List.generate(
+        count,
+        (i) => eg.user(
+          userId: i,
+          email: 'user$i@example.com',
+          fullName: 'User $i'));
+    }
+
+    test('RealmUserAddEvent', () async {
+      final users = generateUsers(count: 1500);
+
+      Future<void> act(PerAccountStore store) async {
+        await store.addUser(eg.user(
+          userId: 10000,
+          email: 'user10000@example.com',
+          fullName: 'User 10000'));
+      }
+
+      void expect(Iterable<MentionAutocompleteResult> results) {
+        check(results).single
         .isA<UserMentionAutocompleteResult>()
         .userId.equals(10000);
-    }
+      }
+
+      await doCheck(
+        users: users,
+        rawQuery: 'User 10000',
+        act: act,
+        expect: expect,
+      );
+    });
+
+    test('RealmUserRemoveEvent', () async {
+      final users = generateUsers(count: 1500);
+
+      Future<void> act(PerAccountStore store) async {
+        await store.removeUser(1495);
+      }
+
+      void expect(Iterable<MentionAutocompleteResult> results) {
+        check(results).isEmpty();
+      }
+
+      await doCheck(
+        users: users,
+        rawQuery: 'User 1495',
+        act: act,
+        expect: expect,
+      );
+    });
+
+    test('RealmUserUpdateEvent', () async {
+      final users = generateUsers(count: 1500);
+
+      Future<void> act(PerAccountStore store) async {
+        await store.updateUser(1, fullName: 'Updated User 1');
+      }
+
+      void expect(Iterable<MentionAutocompleteResult> results) {
+        check(results).single
+        .isA<UserMentionAutocompleteResult>()
+        .userId.equals(1);
+      }
+
+      await doCheck(
+        users: users,
+        rawQuery: 'Updated User 1',
+        act: act,
+        expect: expect,
+      );
+    });
+
+    test('MessageEvent', () async {
+      final users = generateUsers(count: 1500);
+
+      Future<void> act(PerAccountStore store) async {
+        await store.addMessage(eg.dmMessage(from: eg.selfUser, to: [users[100]]));
+      }
+
+      void expect(Iterable<MentionAutocompleteResult> results) {
+        check(results.elementAt(0))
+        .isA<UserMentionAutocompleteResult>()
+        .userId.equals(100);
+      }
+
+      await doCheck(
+        narrow: DmNarrow.withUser(100, selfUserId: eg.selfUser.userId),
+        users: users,
+        rawQuery: '',
+        act: act,
+        expect: expect,
+      );
+    });
+
+    // test('fetching older messages', () async {
+    //   final users = generateUsers(count: 1500);
+    //   final store = eg.store();
+    //   final stream = eg.stream();
+    //   final narrow = StreamNarrow(stream.streamId);
+    //   final subscription = eg.subscription(stream);
+    //   await store.addStream(stream);
+    //   await store.addSubscription(subscription);
```

Comment:
@gnprice I have commented out this test for now as it wouldn't pass. I spent a significant amount of time to diagnose what was the problem, but unfortunately, I wasn't able to figure it out. I would appreciate it if you can have a look at it.

Reply:
Yeah. I think fundamentally the issue here is that these tests using `await Future(() {});` are fragile.

They're trying to make several different things happen in a certain order, but those things are asynchronous in the code. The `await Future(() {})` basically allows each thread of async control flow to advance from one `await` to the next `await`… but what that means in terms of actual behavior of the models depends in a very brittle way on the exact details of how the code being tested is written. Small differences in that code will cause things in the test to happen in a different order and break the test.

That's also why the existing tests of this flavor have so many checks like `check(done).isFalse()` — those are there to make sure that the test's assumptions about what order things happen in are satisfied. That way, if those assumptions are broken by a refactor or by a change to the test, the test fails instead of silently ceasing to successfully test what it's meant to test.

---

It looks like this test is failing at the `check(done).isFalse()` here:
```dart
      await act(store);

      await Future(() {});
      check(done).isFalse();
      await Future(() {});
      check(done).isTrue();
      expect(view.results);
```

so IOW the search is done sooner than the test expects. Presumably it's due to some difference between the `fetchOlder` in this test, and the variations on `handleEvent` found in the neighboring tests.

---

The test passes if you make this change:
```diff
@@ -293,7 +296,6 @@ void main() {
       view.addListener(() { done = true; });
       view.query = MentionAutocompleteQuery(rawQuery);
 
-      await Future(() {});
       check(done).isFalse();
 
       await act(store);
```

But in that case is the test still successfully testing what it's supposed to test? I'm not sure what the assumptions are of your `doCheck` method.

(Probably a chat thread would be a good place to go into further detail.)